src/main.cpp

```diff
@@ -3349,282 +3363,341 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
-        vRereadCPIDs.insert(pindex->GetCPID());
-    }
+        if(pindexBest->IsUserCPID())
+        {
+            // remeber the cpid to re-read later
+            vRereadCPIDs.insert(pindexBest->GetCPID());
+            // The user has no longer staked this block.
+            RemoveCPIDBlockHash(pindexBest->GetCPID(), pindexBest->GetBlockHash());
+        }
 
-    // Re-read researchers history after all blocks disconnected
-    for( const string& sRereadCPID : vRereadCPIDs )
-    {
-        StructCPID stCPID = GetLifetimeCPID(sRereadCPID,"DisconnectBlock()");
-        (void)stCPID;
+        // New best block
+        cnt_dis++;
+        pindexBest = pindexBest->pprev;
+        hashBestChain = pindexBest->GetBlockHash();
+        blockFinder.Reset();
+        nBestHeight = pindexBest->nHeight;
+        nBestChainTrust = pindexBest->nChainTrust;
+
+        // Tally research averages.
+        if (nBestHeight==disconnect_tally_height)
+        {
+            if (fDebug) printf("DisconnectBlocksBatch: TallyNetworkAverages (v9P %%%d) height %d\n",TALLY_GRANULARITY,nBestHeight);
+            TallyNetworkAverages_v9();
+        }
     }
 
-    // Connect longer branch
-    vector<CTransaction> vDelete;
-    for (unsigned int i = 0; i < vConnect.size(); i++)
+    /* fix up after disconnecting, prepare for new blocks */
+    if(cnt_dis>0)
     {
-        CBlockIndex* pindex = vConnect[i];
-        CBlock block;
-        if (!block.ReadFromDisk(pindex))
-            return error("Reorganize() : ReadFromDisk for connect failed");
-        if (!block.ConnectBlock(txdb, pindex, false, true))
+
+        //Block was disconnected - User is Re-eligibile for staking
+        StructCPID sMag = GetInitializedStructCPID2(GlobalCPUMiningCPID.cpid,mvMagnitudes);
+        nLastBlockSolved = 0;
+        if (sMag.initialized)
         {
-            // Invalid block
-            return error("Reorganize() : ConnectBlock %s failed", pindex->GetBlockHash().ToString().substr(0,20).c_str());
+            sMag.LastPaymentTime = 0;
+            mvMagnitudes[GlobalCPUMiningCPID.cpid]=sMag;
         }
 
-        // Queue memory transactions to delete
-        for( const CTransaction& tx : block.vtx )
-            vDelete.push_back(tx);
-    }
+        // Resurrect memory transactions that were in the disconnected branch
+        for( CTransaction& tx : vResurrect)
+            AcceptToMemoryPool(mempool, tx, NULL);
 
-    if (!txdb.WriteHashBestChain(pindexNew->GetBlockHash()))
-        return error("Reorganize() : WriteHashBestChain failed");
+        if (!txdb.TxnCommit())
+            return error("DisconnectBlocksBatch: TxnCommit failed"); /*fatal*/
 
-    // Make sure it's successfully written to disk before changing memory structure
-    if (!txdb.TxnCommit())
-        return error("Reorganize() : TxnCommit failed");
+        // Need to reload all contracts
+        std::string admin_messages;
+        LoadAdminMessages(true, admin_messages);
```

Comment:
I think the order we want is:
- LoadAdminMessages
- Tally
- GetLifetimeCPID

Tally updates the blocks staked by a CPID and that set dictates which blocks to accumulate interest/research rewards from.